prow/crier/controller.go

```diff
@@ -216,6 +216,10 @@ func (c *Controller) processNextItem() bool {
 		return true
 	}
 
+	// start actual reporter logic
+	c.Lock()
```

Comment:
two gerrit jobs can finish and try to report at the same time, and will result in posting multiple report messages

Reply:
How about only locking if we need to report, so something like:

```
if !c.reporter.ShouldReport(pj) {
  c.q.Forget(key); return true
}
c.Lock()
defer c.Unlock()
if !c.reporter.ShouldReport(pj) { c.queue.Forget(key); return true }

if pj.Status.PrevReportStates == nil {
...
```

aka in the normal case (no reporting required) we don't lock.

Reply:
Because we don't need to lock unless we need to report. We should concurrently determine that builds do not need a report.

An in between solution would be to use https://golang.org/pkg/sync/#RWMutex

You can `RLock()` to determine `c.reporter.ShouldReport(pj)` and only `Lock()` (after first doing an `RUnlock()`) if you need to report.

This still unnecessarily blocks PR A whenever PR B is reporting, but it is better than nothing

Reply:
hummm the state we read from `ShouldReport` is depend on the status update after the actual `Report` call though. (Since it's only for gerrit reporter, I'm even fine with restrict gerrit reporter worker to always 1)

Reply:
if it is 1 do we need the locking?

This lock is entirely too broad right now. You need to do something to shrink the critical path.

Reply:
dropped locking for now